### PR TITLE
A new approach to handle the service instance refresh.

### PR DESCRIPTION
**Pull request by cescoffier**
This PR proposes a new approach to handle the service instance "refresh". It reuses the idea of the timed cache we had before but uses the wonderful `memoize` operator that Julien did. It removes a lot of the code that was required before and simplifies how the unis are created.

CC @michalszynkiewicz @jponge

### CODE REVIEW

File: api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java [hunk 1]
```diff
@@ -45,73 +41,26 @@ public CachingServiceDiscovery(ServiceDiscoveryConfig config) {
                     period,
                     e);
         }
+
+        this.lastResults = Collections.emptyList();
+        this.instances = Uni.createFrom().deferred(() -> fetchNewServiceInstances(this.lastResults)
+                .invoke(l -> this.lastResults = l)
+                .onFailure().invoke(this::handleFetchError)
+                .onFailure().recoverWithItem(this.lastResults))
+                .memoize().atLeast(this.refreshPeriod);
```
Comment (michalszynkiewicz) [inline api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: does atLeast literally mean at least? What does it mean in practice?

**Comment by michalszynkiewicz** [comment]
nice!

File: api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java (hunk 1 above)
Comment (jponge) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: It memoizes for at least the given amount of time, then invalidates.

Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: `atLeast` is misleading for me then. But the method is a great fit for what we need :)

Comment (jponge) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: Would `duringAtLeast` have worked better?

Comment (jponge) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: I mean, `Uni::memoize()` is still marked as an experimental API so... 🤣

Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: I read `at least` for `time ∈ [this, +∞)`.
What about `forPeriod`?

Comment (jponge) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: @michalszynkiewicz let's move to a Mutiny discussion: https://github.com/smallrye/smallrye-mutiny/discussions/categories/ideas

Comment (michalszynkiewicz) [reply api/src/main/java/io/smallrye/stork/CachingServiceDiscovery.java:50]: done

**Pull request merged**